add section for spurious loss events (#32)
NTAP/rfc8312bis: draft-eggert-tcpm-rfc8312bis.md

K is set to 0, and W_max is set to the congestion window size at the
beginning of the current congestion avoidance.

## Spurious Loss events

For the case where CUBIC reduces its congestion window in response
to detection of packet loss via duplicate ACKs or timeout, there is a
possibility that the missing ACK would arrive after the congestion
window reduction and the corresponding packet retransmission. For
example, packet reordering which is common in networks could trigger
this behavior. A high degree of packet reordering could cause multiple
events of congestion window reduction where spurious losses are
incorrectly interpreted as congestion signals, thus degrading CUBIC's
performance significantly.

When there is a loss event, A CUBIC implementation SHOULD save the current
value of the following variables before the congestion window reduction.

~~~
    prior_cwnd = cwnd
    prior_ssthresh = ssthresh
    prior_W_max = W_max
    prior_K = K
    prior_epoch_start = epoch_start
~~~

> CUBIC MAY choose the Eifel detection algorithm {{!RFC3522}} or Forward

Reviewer: IIRC Eifel comes with an IPR declaration that makes it problematic for many implementations. I'd cite it as OPTIONAL, with an informative reference.

Author: How do I cite something as OPTIONAL?

Reviewer: You cite it informatively and say that it MAY be used, as you currently are doing.

> RTO-Recovery {{!RFC5682}} or use DSACK {{!RFC3708}} to detect spurious

Reviewer: Should we recommend one of these over the other (i.e., with a SHOULD)?

Author: I used MAY as an implementation could completely do something else (for example, QUIC doesn't require any of these).
But I don't have enough experience to decide what would be better. Please let me know if SHOULD would be ok in case an implementation decides to do something completely different?

Reviewer: I'd at least reorder them so Eifel is last, and whatever we prefer is first. And make all the references informative (`?` instead of `!`).

Author: I rearranged the references and updated them to be informative.

> retransmissions. Once a spurious loss event is detected, CUBIC should

Reviewer: Do we mean that CUBIC MUST choose one of the three, and any of the three MAY be chosen? Or do we mean that CUBIC MAY choose one of the three but also something else?

Author: It's the last one - `Or do we mean that CUBIC MAY choose one of the three but also something else?`

Should I phrase it differently?

Reviewer: ```suggestion
CUBIC MAY implement an algorithm to detect spurious retransmissions,
such as DSACK {{?RFC3708}}, Forward RTO-Recovery {{?RFC5682}}
or Eifel {{?RFC3522}}. Once a spurious loss event is detected, CUBIC should
```

Author: done

restore the original values of above mentioned variables as follows.

~~~

>    cwnd = max(cwnd, prior_cwnd)

Reviewer: Why is this a `max` when the others are not?

Author: @lisongxu had suggested this; I think it is to make sure if the congestion window had increased more than the prior_cwnd before the spurious retransmit is detected, we take the higher value.
Would like to know what @lisongxu thinks about this.

Reviewer: We should explain this, since this doesn't do what the text says (which is "restore the original values").

Author: will wait for @lisongxu comment on this.

Reviewer 2: At an undo, the current cwnd may or may not be lower than the saved prior_cwnd. If the current cwnd is already higher than the saved prior_cwnd, we do not need to undo. 

How about?

_When there is a loss event, A CUBIC implementation SHOULD save the current
value of the following variables before the congestion window reduction._

~~~
    prior_cwnd = cwnd
    prior_ssthresh = ssthresh
    prior_W_max = W_max
    prior_K = K
    prior_epoch_start = epoch_start
    prior_W_est = W_est
~~~

_..................  Once a spurious loss event is detected, CUBIC should restore the original values of the above mentioned variables as follows if the current cwnd is lower than prior_cwnd._

~~~
    if (cwnd < prior_cwnd) {
        cwnd = prior_cwnd
        ssthresh = prior_ssthresh
        W_max = prior_W_max
        K = prior_K
        epoch_start = prior_epoch_start
        W_est = prior_W_est
    }
~~~

Author: Ah.. I am not sure about this one, esp. the variables W_max, epoch_start and K as these dictate the CUBIC curve. For example, W_max for the spurious loss event will be lower than prior_W_max and hence CUBIC will enter Convex phase earlier.
Spurious loss detection should not usually take more than a couple of RTTs and I think the safest thing to do here is to revert all state to previous state, i.e.
```
cwnd = prior_cwnd
        ssthresh = prior_ssthresh
        W_max = prior_W_max
        K = prior_K
        epoch_start = prior_epoch_start
        W_est = prior_W_est

Reviewer 2: Since the goal of the undo is to improve the performance of Cubic, it is reasonable to keep the current cwnd if it is already higher than prior_cwnd. This is why I like to restore cwnd only if cwnd<prior_cwnd. By the way, this is how current Linux TCP does.

But I am not sure whether we should restore other variables. Both are fine with me. Thanks.

Author: ok, I added a special note for congestion window. Please take a look.

Reviewer 2: Hi Vidhi, I thought about it last night, and now I prefer the if-then method as explained below. 

There are two possible scenarios for undo. 

First (usual cases), an undo happens soon after a spurious loss event, and the current cwnd at undo is still lower than prior_cwnd. In these cases, it makes sense to restore the previous values of all the variables (better performance and safe). 

Second (unusual cases), an undo happens long after a spurious loss event, and as a result, the current cwnd at undo is already higher than prior_cwnd. In these cases, it makes sense to keep the current values of all the variables, because their prior values were measured and calculated too long ago and thus may not be good or accurate measurements of the current network condition (bandwidth, delay, ...)

Therefore, the undo code is
~~~
    if (cwnd < prior_cwnd) {
        cwnd = prior_cwnd
        ssthresh = prior_ssthresh
        W_max = prior_W_max
        K = prior_K
        epoch_start = prior_epoch_start
        W_est = prior_W_est
    }
~~~

How do you think?

Thanks!

Author: Sounds good. Will update the PR.

Author: Updated the PR. @lisongxu does this look ok?

Reviewer 2: Thanks, Vidhi! How about replacing "In rare cases, when the detection happens long after a spurious loss event, " with "In rare cases, when the detection happens long after a spurious loss event and the current cwnd is already higher than the prior_cwnd,"?

Thanks!

Author: sounds good. Updated it. Please review, thanks.
